Dive media: Extract thumbnails from videos with ffmpeg

bstoeger opened the pull request:
Signed-off-by: Berthold Stoeger <bstoeger@mail.tuwien.ac.at>

<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
This is a test-commit for @atdotde: Rudimentary video-thumbnail extraction with ffmpeg.

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->
Extract thumbnails from videos via ffmpeg.

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->
Not yet - test-PR
### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->
Not yet - test-PR

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@atdotde

sfuchs79 commented:
I tested this and it works nicely here for me under Windows 10.

Simply out of curiosity I even used my "own" cross compiled ffmpeg (version 3.4.2) from MXE. I had to copy ffmpeg.exe and a couple of DLLs to the Subsurface root installation folder.
Just to create no misunderstanding: This is no vote from my side to put this stuff into the Subsurface Windows installer. I can understand that this might not be the preferred solution. In fact it is ~35MB of additional "stuff". Just wanted to tell you that it worked for me in that way :-)

dirkhh commented:
We can of course *document* how to do this. Here's where you can download things. Here's where you need to copy them.

bstoeger commented:
> I had to copy ffmpeg.exe and a couple of DLLs to the Subsurface root installation folder.

Cool, then maybe for now we don't even have to use a preferences option. Just tell the user where to put extract the files. For Linux it's simply a matter of having ffmpeg in the PATH, which every sensible distribution should be able to do. For Mac: no idea..!?

Since ffmpeg is GPL + LGPL, I think I could host such an archive.

Note that this is WIP. I want to change the format of the thumbnail files such that for videos we also save the duration (if we know it - see #1494). And it should also be able to contain multiple images (every n-seconds of the video).

sfuchs79 commented:
It may get even better. I'm currently playing with ffmpeg configure options to get a tiny and statically linked ffmpeg.exe which is optimized to only do the specific job we want it to do. Unfortunately with my first try I may have removed too many options - it doesn't work any more (but it is only ~5MB <g>). Could you change your PR in a way that I can see the ffmpeg error message? Currently I only get a "Failed reading ffmpeg output".

bstoeger commented:
Changed it to print `stderr` in case of failure. The code to print `stdin` is commented out. Perhaps you want to change this to a `printf` so that it is properly formatted. We could also show a `QMessageBox`, perhaps of the collected failures after addition of a bunch of videos.

sfuchs79 commented:
Thanks! With help of the ffmpeg error messages I cross compiled ffmpeg with a few other configurations. I now ended up with a statically linked ffmpeg.exe of about 10MB which does the job of creating thumbnails for me.

Config was (copied from MXE ffmpeg.mk):
         --cross-prefix='$(TARGET)'- \
        --enable-cross-compile \
        --arch=$(firstword $(subst -, ,$(TARGET))) \
        --target-os=mingw32 \
        --prefix='$(PREFIX)/$(TARGET)' \
        --enable-static --disable-shared \
        --enable-small \
        --yasmexe='$(TARGET)-yasm' \
        --disable-debug \
        --disable-pthreads \
        --enable-w32threads \
        --disable-doc \
        --disable-ffplay \
        --disable-ffprobe \
        --disable-ffserver \
        --disable-bsfs \
        --disable-encoders \
        --enable-encoder=mjpeg \
        --disable-filters \
        --enable-filter=scale \
        --disable-muxers \
        --enable-muxer=image2

atdotde commented:
Of course, you can also use ffmpeg to figure out the duration (or rather its helper ffprobe):

https://superuser.com/questions/650291/how-to-get-video-duration-in-seconds

I don't know if shipping ffmpeg is such a great idea. Why not rather relay on it being there or directing the users to 

https://www.ffmpeg.org/download.html

or using the package manager of their system (on Mac, home-brew happily installs it and places it in the path).

bstoeger commented:
> Of course, you can also use ffmpeg to figure out the duration (or rather its helper ffprobe)

Yes, but we should strive to give the user the best possible experience without having to install external software as long a the effort stays within reasonable limits. Extracting date and especially duration turns out to be very simple (I'm currently looking at .wmv). I don't want to make ffmpeg a dependency for all video-manipulations. Only for thumbnailing.

bstoeger commented:
Rebased on master.
Added three preferences fields, which control extraction (quote from commit-message):
- `extract_video_thumbnails` (bool): if true, thumbnails are calculated.
- `extract_video_thumbnail_position` (int 0..100): position in video where thumbnail is fetched.
- `ffmpeg_executable (string)`: path of ffmpeg executable.

If ffmpeg fails to start, thumbnailing is turned off. Preferences-fields not yet wired to frontend!

bstoeger commented:
Added a "video-file-marker" to mark video thumbnails. The "design" is pure garbage - this will have to be fixed by a more capable person (file: `icons/video_overlay.svg`). Note that the marker is added on-the-fly. The thumbnails on-disk are not messed up. I.e. nothing should be destroyed by the PR as-is.

bstoeger commented:
Update:
- connected the preferences values to the UI.
- render duration on thumbnails in dive-photo-tab.

@dirkhh, @sfuchs79, @atdotde: From my side this is feature complete, therefore I allowed my self to mark the last commit as fixes #359. This still needs some UI / graphics polishing and there are many options for improvement (@atdotde has the idea of extracting frames every n-seconds; we could also snatch half-a-second of movie, like youtube does). But personally, I'd like to limit my work on this to bug-fixing and polishing, so that I can bother with other things, notably undo.

bstoeger commented:
I replaced the work-in-progress by the needs-testing label, because from my side, this is ready. 

I would really prefer if this was addressed before the corresponding preferences-code is reworked. Doing it once was painful enough.

janiversen commented:
If you tell which parts of prefs it is, I will hold thst part back a while.

bstoeger commented:
```
bool GeneralSettingsObjectWrapper::extractVideoThumbnails() const
{
	return prefs.extract_video_thumbnails;
}
 int GeneralSettingsObjectWrapper::extractVideoThumbnailsPosition() const
{
	return prefs.extract_video_thumbnails_position;
}
 QString GeneralSettingsObjectWrapper::ffmpegExecutable() const
{
	return prefs.ffmpeg_executable;
}
```

That would be very nice - thanks.

dirkhh commented on desktop-widgets/preferences/preferences_defaults.ui:
```diff
@@ -6,8 +6,8 @@
-    <width>561</width>
-    <height>558</height>
+    <width>588</width>
+    <height>600</height>
```
this always worries me - I assume it's just automatic tool output, but we used to struggle a lot with the preferences window suddenly not fitting in our design goal screen size of 1024x600 and people not being able to change preferences

bstoeger replied:
Yes - that was my fault. Apparently I played with the window-size in Qt Creator. I wonder: Couldn't the window-size be calculated from the preferred sizes of the widgets...?

I removed the gratuitous `leftMargin`, etc. changes. It's really rude of Qt Creator to mangle these things.

bstoeger commented:
> Which brings me to what this really needs.
> Testing on Windows, Mac, and Linux.

I tested it quite extensively on Kubuntu. Of course, experience shows that testers who are not the author often find problems in use-cases that the author never thought about.

@sfuchs79 tested an older version on Windows.

I don't have access to Mac. So out in that respect. 

> Oh, and an addition to the user manual :-)

Yes, looked into it a few days ago and quickly gave up. There's quite a lot to change!

dirkhh commented:
Tested on Mac, both with and without `ffmpeg`
VERY nice work @bstoeger
